fcronsighup.c

```diff
@@ -86,35 +86,44 @@ sig_daemon(void)
     /* SIGHUP is sent once 10s before the next minute to avoid
      * some bad users to block daemon by sending it SIGHUP all the time */
 {
-    /* we don't need to make root wait */
-    if (uid != rootuid) {
-        time_t t = 0;
-        int sl = 0;
+    int max_delay_s = 60;
+#ifdef MAX_FCRONTAB_LOAD_DELAY_SECONDS
+    max_delay_s = MAX_FCRONTAB_LOAD_DELAY_SECONDS;
+#endif
+    if (uid == rootuid) {
+      /* we don't need to make root wait */
+      max_delay_s = 0;
+    }
+
+    if (max_delay_s > 0) {
+        time_t now_s = 0;
+        int delay_s = 0;
         FILE *fp = NULL;
         int fd = 0;
         struct tm *tm = NULL;
         char sigfile[PATH_LEN];
         char buf[PATH_LEN];
 
         sigfile[0] = '\0';
-        t = time(NULL);
-        tm = localtime(&t);
-
-        if ((sl = 60 - (t % 60) - 10) < 0) {
-            if ((tm->tm_min = tm->tm_min + 2) >= 60) {
-                tm->tm_hour++;
-                tm->tm_min -= 60;
-            }
-            snprintf(buf, sizeof(buf), "%02d:%02d", tm->tm_hour, tm->tm_min);
-            sl = 60 - (t % 60) + 50;
+        now_s = time(NULL);
+
+        /* target 10s before the end of the current minute */
+        delay_s = 50 - (now_s % 60);
+        if (delay_s < 0) {
+          /* target 10 before the end of the next minute */
+          delay_s += 60;
         }
-        else {
-            if (++tm->tm_min >= 60) {
-                tm->tm_hour++;
-                tm->tm_min -= 60;
-            }
-            snprintf(buf, sizeof(buf), "%02d:%02d", tm->tm_hour, tm->tm_min);
+
+        if (delay_s > max_delay_s) {
+          delay_s = max_delay_s;
         }
+
+        /* tm is now + delay */
+        tm = localtime(&now_s);
+        tm->tm_sec += delay_s;
+        mktime(tm);
```

Comment:
first: yea, your proposal is obviously a little nicer (no discarded retval, for one thing). ill apply it when i fix any other comments.

but: re-normalization is one of the most common uses of mktime(), and this behavior has worked for at least 30 years.

the specific rule that all non-derived time fields are explicitly permitted be outside of their ranges specified exists since POSIX.1-1988. you can find it in ansi c, iso c, anywhere.

```
The original values of the tm_wday and tm_yday components of the structure shall be ignored, and the original values of the other components shall not be restricted to the ranges described in [<time.h>].
```
https://pubs.opengroup.org/onlinepubs/9699919799/



it is more explicitly in linux libc ctime(3) mktime(3)
```
if structure members are outside their valid interval, they will be normalized (so that, for example, 40 October is changed into 9 November);
```

Reply:
refactored to remove relying on side effects of mktime, and clean up a little bit since i had to retest anyway